File: projects/test_models/CSharpModel/Types/Package2/ChildViaSettings.cs
// -*- mode: csharp; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
//
// Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
// MA 02110-1301, USA.
//
using System;

namespace Dogen.TestModels.CSharpModel.Package2
{
    public sealed class ChildViaSettings : Dogen.TestModels.CSharpModel.Package2.Parent
    {
        #region Properties
        public int prop_1 { get; set; }
        #endregion

        #region Constructors
        public ChildViaSettings() { }
        #endregion

        #region Equality
        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;

            var value = obj as ChildViaSettings;
            if (value == null || !base.Equals(value)) return false;

            return
                prop_1 == value.prop_1;
        }

        public static bool operator ==(ChildViaSettings lhs, ChildViaSettings rhs)
        {
            if (Object.ReferenceEquals(lhs, rhs))
                return true;

            return !Object.ReferenceEquals(null, lhs) && lhs.Equals(rhs);
        }

        public static bool operator !=(ChildViaSettings lhs, ChildViaSettings rhs)
        {
            return !(lhs == rhs);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                // Choose large primes to avoid hashing collisions
                const int HashingBase = (int) 2166136261;
                const int HashingMultiplier = 16777619;

                int hash = HashingBase;
                hash = (hash * HashingMultiplier) ^ prop_1.GetHashCode();
                return hash;
            }
        }
        #endregion

        #region Dumpers
        internal override string Dump()
        {
            return ChildViaSettingsDumper.Dump(this);
        }
        #endregion
    }
}

File: projects/test_models/cpp_model/src/types/package_8/child_via_settings.cpp
/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
 *
 * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 *
 */
#include <ostream>
#include "dogen/test_models/cpp_model/io/package_7/parent_io.hpp"
#include "dogen/test_models/cpp_model/types/package_8/child_via_settings.hpp"

namespace dogen {
namespace test_models {
namespace cpp_model {
namespace package_8 {

child_via_settings::child_via_settings()
    : prop_1_(static_cast<int>(0)) { }

child_via_settings::child_via_settings(
    const int prop_0,
    const int prop_1)
    : dogen::test_models::cpp_model::package_7::parent(prop_0),
      prop_1_(prop_1) { }

void child_via_settings::to_stream(std::ostream& s) const {
    s << " { "
      << "\"__type__\": " << "\"dogen::test_models::cpp_model::package_8::child_via_settings\"" << ", "
      << "\"__parent_0__\": ";
    dogen::test_models::cpp_model::package_7::parent::to_stream(s);
    s << ", "
      << "\"prop_1\": " << prop_1_
      << " }";
}

void child_via_settings::swap(child_via_settings& other) noexcept {
    dogen::test_models::cpp_model::package_7::parent::swap(other);

    using std::swap;
    swap(prop_1_, other.prop_1_);
}

bool child_via_settings::equals(const dogen::test_models::cpp_model::package_7::parent& other) const {
    const child_via_settings* const p(dynamic_cast<const child_via_settings* const>(&other));
    if (!p) return false;
    return *this == *p;
}

bool child_via_settings::operator==(const child_via_settings& rhs) const {
    return dogen::test_models::cpp_model::package_7::parent::compare(rhs) &&
        prop_1_ == rhs.prop_1_;
}

child_via_settings& child_via_settings::operator=(child_via_settings other) {
    using std::swap;
    swap(*this, other);
    return *this;
}

int child_via_settings::prop_1() const {
    return prop_1_;
}

void child_via_settings::prop_1(const int v) {
    prop_1_ = v;
}

} } } }
